pkg/clients/compute/agentpool/agentpool.go

```diff
@@ -0,0 +1,152 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package agentpool
+
+import (
+	"github.com/crossplane/provider-azure/pkg/clients/compute/consts"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-03-01/containerservice"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/crossplane/provider-azure/apis/compute/v1alpha3"
+	azure "github.com/crossplane/provider-azure/pkg/clients"
+)
+
+// New returns an Azure AgentPool object from a AgentPool spec
+func New(c *v1alpha3.AgentPool) containerservice.AgentPool {
+	p := containerservice.AgentPool{
+		ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
+			VMSize:            containerservice.VMSizeTypes(c.Spec.NodeVMSize),
+			EnableAutoScaling: azure.ToBoolPtr(c.Spec.EnableAutoScaling),
+			// AgentPool APIs supported only for VMSS agentpools.
+			// For more information, please check https://aka.ms/multiagentpoollimitations
+			Type: containerservice.VirtualMachineScaleSets,
+			// Defaults override below if needed
+			Mode: containerservice.User,
+		},
+	}
+	if c.Spec.Mode != "" {
+		p.ManagedClusterAgentPoolProfileProperties.Mode = containerservice.AgentPoolMode(c.Spec.Mode)
+	}
+	if len(c.Spec.AvailabilityZones) > 0 {
+		p.AvailabilityZones = &c.Spec.AvailabilityZones
+	}
+	if c.Spec.NodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.Count = c.Spec.NodeCount
+	}
+	if c.Spec.MinNodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.MinCount = c.Spec.MinNodeCount
+	}
+	if c.Spec.MaxNodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.MaxCount = c.Spec.MaxNodeCount
+	}
+	if c.Spec.VnetSubnetID != "" {
+		p.VnetSubnetID = to.StringPtr(c.Spec.VnetSubnetID)
+	}
+	if len(c.Spec.NodeTaints) > 0 {
+		p.NodeTaints = &c.Spec.NodeTaints
+	}
+	return p
+}
+
+// NewProfile returns an Azure ManagedClusterAgentPoolProfile object from a AKSCluster spec
+func NewProfile(c *v1alpha3.AKSCluster) containerservice.ManagedClusterAgentPoolProfile {
+	nodeCount := int32(v1alpha3.DefaultNodeCount)
+	if c.Spec.NodeCount != nil {
+		nodeCount = int32(*c.Spec.NodeCount)
+	}
+	p := containerservice.ManagedClusterAgentPoolProfile{
+		Name:   to.StringPtr(consts.AgentPoolProfileName),
+		Count:  &nodeCount,
+		VMSize: containerservice.VMSizeTypes(c.Spec.NodeVMSize),
+		Mode:   containerservice.System,
+		Type:   containerservice.VirtualMachineScaleSets,
+	}
+	if c.Spec.VnetSubnetID != "" {
+		p.VnetSubnetID = to.StringPtr(c.Spec.VnetSubnetID)
+	}
+	return p
+}
+
+// NeedUpdate determines if a AgentPool need to be updated
+func NeedUpdate(c *v1alpha3.AgentPool, az *containerservice.AgentPool) bool {
+	if c.Spec.NodeVMSize != string(az.VMSize) {
+		return true
+	}
+	if nodeCountNeedUpdate(c, az) {
+		return true
+	}
+	if c.Spec.VnetSubnetID != "" {
```

Comment:
What if `spec.vnetSubnetID` is "" but `az.VnetSubnetID != nil`, e.g., infrastructure object has been modified out of band? We need to be able to reconcile it back to the state specified by the `spec`.

Reply:
I've marked `vnetSubnetID` as immutable